async/delay.ts

```diff
@@ -39,12 +39,20 @@ export interface DelayOptions {
 export function delay(ms: number, options: DelayOptions = {}): Promise<void> {
   const { signal, persistent } = options;
   if (signal?.aborted) {
-    return Promise.reject(new DOMException("Delay was aborted.", "AbortError"));
+    return Promise.reject(
+      typeof signal.reason === "undefined"
```

Comment:
I think this can be condensed to `signal.reason ?? 'Delay was aborted.'` but in principal this change looks like a no brainier to me.

Reply:
Thames for the feedback! I considered that, but then realized it would prevent `null` from being forwarded, and it's just as meaningful a value as any other.

Reply:
It looks like that `signal.reason` can't be `undefined` when it's aborted.

```
> let ac = new AbortController()
undefined
> ac.abort(undefined)
undefined
> ac.signal.reason
DOMException: The signal has been aborted
```

, which means `new DOMException("Delay was aborted.", "AbortError")` part is never reached.

How about checking if `reason` is DOMException and AbortError, and returning the default error in that case (assuming that the user didn't give `reason` in that case)

Reply:
@kt3k

> It looks like that `signal.reason` can't be `undefined` when it's aborted.
> 
> ```
> > let ac = new AbortController()
> undefined
> > ac.abort(undefined)
> undefined
> > ac.signal.reason
> DOMException: The signal has been aborted
> ```
> 
> , which means `new DOMException("Delay was aborted.", "AbortError")` part is never reached.

According to the [spec](https://dom.spec.whatwg.org/#interface-AbortSignal), that's right — and Deno's implementation appears to conform in that way. The only reason I special-cased `undefined` was as a grace measure for a potential case where a nonstandard `AbortSignal` is provided (one that doesn't create a value when no argument is provided to the `abort` method) — this module is marked as browser-compatible after all, and Deno isn't the only runtime where it's used.

In hindsight, that might have been overengineering — I reversed those changes, removing the default reason entirely, and updated the tests to clarify the intent of the change — **to reject the exact `reason` value provided by the signal** instead of ignoring it and using another value.

> How about checking if `reason` is DOMException and AbortError, and returning the default error in that case (assuming that the user didn't give `reason` in that case)

In the case that a user provides a custom `DOMException` with the name `"AbortError"`, it seems like quite an arbitrary decision to ignore that value. Is there a way to positively identify Deno's default parameter instance? Something like this wouldn't even be enough AFAICT:

```ts
const r = signal.reason;

// Could still be a user-created value:
const looksLikeDenosDefaultParameter =
  r instanceof DOMException &&
  r.name === "AbortError" &&
  r.message === "The signal has been aborted";
```

Unless Deno brands all of its internally created values with a unique symbol or something like that, I'm not sure how we could discriminate it.

Reply:
Yeah, in fact you should just take it if the reason is any kind of an Error object. And if its not, frankly you should not use the signal reason as the error message either but add it to a field. Such as this logic:

```ts
class AbortError extends DOMException {
  constructor(public readonly reason: unknown) {
    super("The signal has been aborted")
  }
}

const err = signal.reason instanceof Error
  ? signal.reason
  : new AbortError(signal.reason)
```

This will provide an Error object consistently, with a consistent message but will also provide the reason on a field should that be provided by the aborter.


This covers these cases:
```ts
controller.abort() // DOMException: The signal has been aborted
controller.abort('some other reason') // DOMException: The signal has been aborted { reason: 'some other reason' }
controller.abort(null) // DOMException: The signal has been aborted { reason: null }
```

The abort reason can be any value, so I would put it onto a field on the exception rather than try to make it the message.

Reply:
@justinmchase I definitely agree with you that it's best practice to only throw/reject instances of the [`Error`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) class (which includes [`DOMException`](https://developer.mozilla.org/en-US/docs/Web/API/DOMException) instances) — even if the language allows for other types of values.

However, the whole point of this PR is for this function to proxy the user's value verbatim (whether or not they choose to follow that practice). See the newly-added test, specifically [the step in lines 50–57](https://github.com/denoland/deno_std/pull/3479/files#diff-30bc1a4d7912bd1b9ba4c411ccb0589508f5099b3e9dea07035ce48902b38ee1R50-R57) — [`assertStrictEquals`](https://github.com/denoland/deno_std/blob/0.193.0/testing/asserts.ts#L253) uses [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) for identity comparison, so moving the user value to a [property](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause) on a different value would cause that test to fail — and it would work against the goal of this PR, which is primarily to simplify exception-handling for users. By rejecting the reason argument provided to `abortSignal.abort(reason)`, the user can handle this value directly in their `catch` code, without having to learn about yet another abstraction in a function in `std`.

Here's a contrived example, but it illustrates what I described:

```ts
const reason = new Error("That took too long!");

const controller = new AbortController();
delay(500).then(() => controller.abort(reason));

try {
  await delay(1000, { signal: controller.signal });
} catch (cause) {
  if (controller.signal.aborted && Object.is(cause, reason)) {
    console.log("I know that my code caused the exception because:", cause.message);
  }
}
```

If we create a new value (even if we attach the user's reason to it somehow), the user has to learn about that abstraction in order to handle the exception cause: which creates an unnecessary challenge.